test/apps/iodemo/io_demo.cc

```diff
@@ -0,0 +1,456 @@
+    long        iter_count;
+    long        window_size;
+    io_op_t     operation;
+    bool        verbose;
+} options_t;
+
+
+class P2pDemoCommon : protected UcxContext {
+protected:
+    /* IO request header */
+    typedef struct {
+        io_op_t     op;
+        uint32_t    sn;
+        size_t      data_size;
+    } iomsg_hdr_t;
+
+    /* Asynchronous IO message */
+    class IoMessage : public UcxCallback {
+    public:
+        IoMessage(size_t buffer_size, io_op_t op, uint32_t sn, size_t data_size) :
+            _buffer(malloc(buffer_size)) {
+            iomsg_hdr_t *hdr = reinterpret_cast<iomsg_hdr_t*>(_buffer);
+            assert(sizeof(*hdr) <= buffer_size);
+            hdr->op          = op;
+            hdr->sn          = sn;
+            hdr->data_size   = data_size;
+        }
+
+        ~IoMessage() {
+            free(_buffer);
+        }
+
+        virtual void operator()() {
+            delete this;
+        }
+
+        void *buffer() {
+            return _buffer;
+        }
+
+    private:
+        void *_buffer;
+    };
+
+    P2pDemoCommon(const options_t& test_opts) :
+        UcxContext(test_opts.iomsg_size, test_opts.verbose),
+        _test_opts(test_opts) {
+
+        _data_buffer.resize(opts().data_size + ALIGNMENT);
+        uintptr_t ptr = (uintptr_t)&_data_buffer[0];
+        _padding = ((ptr + ALIGNMENT - 1) & ~(ALIGNMENT - 1)) - ptr;
+    }
+
+    const options_t& opts() const {
+        return _test_opts;
+    }
+
+    void *buffer() {
+        return &_data_buffer[_padding];
+    }
+
+    void send_io_message(UcxConnection *conn, io_op_t op, uint32_t sn) {
+        IoMessage *m = new IoMessage(opts().iomsg_size, op, sn,
+                                     opts().data_size);
+        verbose_os() << "sending IO " << io_op_names[op] << ", sn " << sn
+                     << " data size " << opts().data_size << std::endl;
+        conn->send_io_message(m->buffer(), opts().iomsg_size, m);
+    }
+
+private:
+    const options_t _test_opts;
+    std::string     _data_buffer;
+    size_t          _padding;
+};
+
+
+class DemoServer : private P2pDemoCommon {
+public:
+    // sends an IO response when done
+    class IoWriteResponseCallback : public UcxCallback {
+    public:
+        IoWriteResponseCallback(DemoServer *server, UcxConnection* conn,
+                                uint32_t sn) :
+            _server(server), _conn(conn), _sn(sn) {
+        }
+
+        virtual void operator()() {
+            _server->send_io_message(_conn, IO_COMP, _sn);
+            delete this;
+        }
+
+    private:
+        DemoServer*     _server;
+        UcxConnection*  _conn;
+        uint32_t        _sn;
+    };
+
+    DemoServer(const options_t& test_opts) : P2pDemoCommon(test_opts) {
+    }
+
+    void run() {
+        struct sockaddr_in listen_addr;
+        memset(&listen_addr, 0, sizeof(listen_addr));
+        listen_addr.sin_family      = AF_INET;
+        listen_addr.sin_addr.s_addr = INADDR_ANY;
```

Comment:
Maybe add an argument for sockaddr to listen? If provided use it, otherwise use `INADDR_ANY`

Reply:
sounds like a feature request which could be implemented in separate PR